custom/icds/integrations/interfaces/phi.py

```diff
@@ -0,0 +1,91 @@
+from django.conf import settings
+
+from custom.icds.integrations.clients.phi import PHIApiClient
+from custom.icds.integrations.interfaces.exceptions.phi import (
+    AuthenticationFailure,
+    MissingCredentials,
+)
+
+
+class BasePHIInterface(object):
+    client_type = None
+
+    def __init__(self):
+        self._response = None
+
+    def _ensure_credentials(self):
+        if not self._api_key or not self._password:
+            raise MissingCredentials("Please set api key and password for phi")
+
+    @property
+    def _data(self):
+        raise NotImplementedError
+
+    def send_request(self):
+        self._response = self._client.send_request()
+        if self._response.status_code == 403:
+            raise AuthenticationFailure("Authentication Failed")
+        return self._parse_response()
```

Comment:
`_parse_response()` is temporally coupled to `send_request` - it has a dependency on state from `send_request`.  You can make this dependency explicit by passing `_response` as a parameter to `_parse_response` rather than mutating the class in the middle.

Reply:
i see. i did that earlier.
I did store the raw response on `_response` so that it can be referred later if needed. and then it made sense to just use it instead of passing it to the method. So you are suggesting, i store it on `_response` but still pass it as an argument?

Reply:
I think it'd be better to split it up.  One problem is setting up and making the request, and one is dealing with the result.  That request makes a logical breaking point.  If one class deals with both things, then half of the class only applies to before the request, while the other half only applies to after the request.  You can split it up to one unit of code (IMO a function would be better) to prepare the request, and one to interact with the result.

Think of how `requests` works - you pass the parameters to a function, which does the action and returns a separate "response" object.  That object can be interacted with in multiple ways if necessary.  Right now this is probably simple enough to just return the parsed JSON, but if the needs became more complicated, I could see wrapping that in a custom class which knows how to access various parts of the response.